core-blocks/file/edit.js

```diff
@@ -0,0 +1,228 @@
+			this.setState( { showCopyConfirmation: true } );
+		};
+		const resetCopyConfirmation = () => {
+			this.setState( { showCopyConfirmation: false } );
+		};
+
+		// Choose Media File or Attachment Page (when file is in Media Library)
+		const changeLinkDestinationOption = ( newHref ) => {
+			setAttributes( { textLinkHref: newHref } );
+		};
+		const changeOpenInNewWindow = ( newValue ) => {
+			setAttributes( {
+				openInNewWindow: newValue ? '_blank' : false,
+			} );
+		};
+		const changeShowDownloadButton = ( newValue ) => {
+			setAttributes( { showDownloadButton: newValue } );
+		};
+
+		if ( ! href ) {
+			return (
+				<MediaPlaceholder
+					icon="media-default"
+					labels={ {
+						title: __( 'File' ),
+						name: __( 'a file' ),
+					} }
+					onSelect={ this.onSelectFile }
+					notices={ noticeUI }
+					onError={ noticeOperations.createErrorNotice }
+					accept="*"
+					type="*"
+				/>
+			);
+		}
+
+		return (
+			<Fragment>
+				<FileBlockInspector
+					hrefs={ { href, textLinkHref, attachmentPage } }
+					{ ...{
+						openInNewWindow,
+						showDownloadButton,
+						changeLinkDestinationOption,
+						changeOpenInNewWindow,
+						changeShowDownloadButton,
+					} }
+				/>
+				<BlockControls>
+					<Toolbar>
+						<MediaUpload
+							onSelect={ this.onSelectFile }
+							type="*"
+							value={ id }
+							render={ ( { open } ) => (
+								<IconButton
+									className="components-toolbar__control"
+									label={ __( 'Edit file' ) }
+									onClick={ open }
+									icon="edit"
+								/>
+							) }
+						/>
+					</Toolbar>
+				</BlockControls>
+				<div className={ classNames }>
+					<div>
+						<FileBlockEditableLink
+							className={ className }
+							placeholder={ __( 'Write file name…' ) }
+							text={ fileName }
+							href={ textLinkHref }
+							updateFileName={ ( text ) => setAttributes( { fileName: text } ) }
+						/>
+						{ showDownloadButton &&
+							<div className={ `${ className }__button-richtext-wrapper` }>
+								<RichText
+									tagName="div" // must be block-level or else cursor disappears
+									className={ `${ className }__button` }
+									value={ buttonText }
+									formattingControls={ [] } // disable controls
+									placeholder={ __( 'Add text…' ) }
+									keepPlaceholderOnFocus
+									multiline="false"
+									onChange={ ( text ) => setAttributes( { buttonText: text } ) }
+								/>
+							</div>
+						}
+					</div>
+					{ isSelected &&
+						<ClipboardButton
+							isDefault
+							text={ href }
+							className={ `${ className }__copy-url-button` }
+							onCopy={ confirmCopyURL }
+							onFinishCopy={ resetCopyConfirmation }
+						>
+							{ showCopyConfirmation ? __( 'Copied!' ) : __( 'Copy URL' ) }
+						</ClipboardButton>
+					}
+				</div>
+			</Fragment>
+		);
+	}
+}
+
+export default compose( [
+	withAPIData( ( props ) => ( {
+		media: `/wp/v2/media/${ props.attributes.id }`,
```

Comment:
You should be rather using [getMedia](https://github.com/WordPress/gutenberg/search?q=getMedia&unscoped_q=getMedia) selector instead of `withAPIData` which is going to be deprecated as soon as we find a way to port all existing code that handles REST API requests. This will also remove the need to introducing changes proposed #7108.

Reply:
```js
export default compose( [
    withSelect( ( select, props ) => {
        const { getMedia } = select( 'core' );
        return {
            image: getMedia( props.attributes.id ), // not sure if getMedia handles undefined values
        };
    } ),
    withNotices,
] )( FileEdit );
```

Reply:
Thank you, I wasn't aware of this. I'll look into it.